File: 006_ZigZag_Conversion/006_1.py
from typing import List, Dict
from queue import Queue
from collections import deque, Counter
import math
from heapq import heappush, heappushpop

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if not s or s == "" or numRows < 1:
            return ""
        if numRows == 1:
            return s
        tmp = [[] for _ in range(numRows)]
        for i in range(len(s)):
            ans = i // (numRows - 1)
            cur = i %  (numRows - 1)
            if ans % 2 == 0:
                tmp[cur].append(s[i])
            else:
                tmp[numRows - 1 - cur].append(s[i])
        res = []
        for i in range(numRows):
            res.append("".join(tmp[i]))
        return "".join(res)


if __name__ == "__main__":
    s = Solution()
    str_ = "LEETCODEISHIRING"
    print(s.convert(str_, 4))

File: 006_ZigZag_Conversion/006_1.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.Queue;
import java.util.ArrayDeque;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Stack;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.Comparator;

class Solution {
    public String convert(String s, int numRows) {
        if (numRows < 2) return s;
        List<StringBuilder> rows = new ArrayList<>();
        for (int i = 0; i < numRows; i++) rows.add(new StringBuilder());
        int i = 0, flag = -1;
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            rows.get(i).append(c);
            if (i == 0 || i == numRows - 1) flag = -flag;
            i += flag;
        }
        StringBuilder res = new StringBuilder();
        for (StringBuilder row : rows) res.append(row);
        return res.toString();
    }
}

public class leetcode {
    public static void main(String[] args) {
        Solution s = new Solution();
        String str = "LEETCODEISHIRING";
        System.out.println(s.convert(str, 4));
    }
}


